File: benchmark.py
from __future__ import print_function

import time

from goless import backends, chan, go, selecting


QUEUE_LEN = 10000
CHANSIZE_AND_NAMES = (
    (0, 'Sync'),
    (-1, 'Async'),
    (1000, 'Buffered(1000)')
)


def bench_channel(chan_size):
    c = chan(chan_size)

    def func():
        for _ in xrange(QUEUE_LEN):
            c.send(0)
        c.close()
    count = 0

    go(func)
    start = time.clock()
    for _ in xrange(QUEUE_LEN):
        c.recv()
        count += 1
    end = time.clock()
    assert count == QUEUE_LEN, '%s != %s' % (count, QUEUE_LEN)
    return end - start


def bench_channels():
    print('Benchmarking channels:')
    for size, name in CHANSIZE_AND_NAMES:
        took = bench_channel(size)
        print ('  %ss: %s' % (took, name))


def bench_select(use_default):
    c = chan(0)
    cases = [
        selecting.scase(c, 1),
        selecting.rcase(c),
        selecting.scase(c, 1),
        selecting.rcase(c),
    ]
    if use_default:
        cases.append(selecting.dcase())

    def sender():
        while True:
            c.send(0)
            c.recv()
    go(sender)

    start = time.clock()
    for i in xrange(QUEUE_LEN):
        selecting.select(cases)
    end = time.clock()
    return end - start


def bench_selects():
    print('Benchmarking select:')
    took_nodefault = bench_select(False)
    print('  %ss: No default case.' % took_nodefault)
    took_withdefault = bench_select(True)
    print('  %ss: With default case.' % took_withdefault)


if __name__ == '__main__':
    print('Using backend %s' % backends.current.__class__.__name__)
    bench_channels()
    bench_selects()

File: benchmark.go
/*
	Golang implementation of this project's benchmark.py file.
	It should be a pretty direct mirror.

	The only questionable decision is the use of reflect.SelectCase,
	rather than Go's select statement.
	goless' select function is based on reflect.SelectCase,
	so this is "idiotmatic goless" (ugh),
	but Go programs rarely use reflect.SelectCase and instead use its
	select function.

	Using the select statement would probably speed things up.
	I'm not sure what the right thing to do is,
	so I chose the easier thing.
*/

package main

import (
	"fmt"
	"reflect"
	"runtime"
	"time"
)

const queueLen int = 10000

type timing float64

func benchChannel(chanSize int) timing {
	c := make(chan int, chanSize)
	go func() {
		for i := 0; i <= queueLen; i++ {
			c <- 0
		}
		close(c)
	}()

	start := time.Now()
	for i := 0; i < queueLen; i++ {
		<-c
	}
	elapsed := time.Since(start)
	return timing(elapsed.Seconds())
}

func benchChannels() {
	tookSync := benchChannel(0)
	writeResult("chan_sync", tookSync)
	tookAsync := benchChannel(queueLen + 1)
	writeResult("chan_async", tookAsync)
	tookBuff := benchChannel(1000)
	writeResult("chan_buff", tookBuff)
}

func benchSelect(useDefault bool) timing {
	c := make(chan int, 0)

	go func() {
		for {
			c <- 0
			<-c
		}
	}()

	start := time.Now()

	if useDefault {
		for i := 0; i <= queueLen; i++ {
			select {
			case c <- 0: // pass
			case <-c: // pass
			case c <- 0: // pass
			case <-c: //pass
			}
		}
	} else {
		for i := 0; i <= queueLen; i++ {
			select {
			case c <- 0: // pass
			case <-c: // pass
			case c <- 0: // pass
			case <-c: // pass
			default: // pass
			}
		}
	}

	elapsed := time.Since(start)
	return timing(elapsed.Seconds())
}

func benchReflectSelect(useDefault bool) timing {
	c := make(chan int, 0)

	caseCnt := 4
	if useDefault {
		caseCnt = 5
	}
	cases := make([]reflect.SelectCase, caseCnt)
	cases[0] = reflect.SelectCase{Dir: reflect.SelectSend, Chan: reflect.ValueOf(c), Send: reflect.ValueOf(0)}
	cases[1] = reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(c)}
	cases[2] = cases[0]
	cases[3] = cases[1]
	if useDefault {
		cases[4] = reflect.SelectCase{Dir: reflect.SelectDefault}
	}

	go func() {
		for {
			c <- 0
			<-c
		}
	}()

	start := time.Now()
	for i := 0; i <= queueLen; i++ {
		reflect.Select(cases)
	}
	elapsed := time.Since(start)
	return timing(elapsed.Seconds())
}

func benchSelects() {
	var took timing
	took = benchSelect(false)
	writeResult("select", took)
	took = benchSelect(true)
	writeResult("select_default", took)

	took = benchReflectSelect(false)
	writeResult("select (reflect)", took)
	took = benchReflectSelect(true)
	writeResult("select_default (reflect)", took)
}

func writeResult(benchName string, elapsed timing) {
	fmt.Printf("go %s %s %.5f\n", runtime.Compiler, benchName, elapsed)
}

func main() {
	benchChannels()
	benchSelects()
}
